Synchronize ufal and dataquest main branches

kosarko opened the pull request:
This contains the following fixes:

Issue 1165: Handle PaginantionException in Clarin licences search requests (ufal/clarin-dspace#1184)
Issue 1055: fixed security issue for downloading file with non anonymous license (ufal/clarin-dspace#1188)
Issue 1186: catch DSpaceBadRequestException rather than BadRequestException (ufal/clarin-dspace#1187)
Issue ufal/clarin-dspace#1190 - fixed Server error  (ufal/clarin-dspace#1193)
checkstyle
Shortref v7 (ufal/clarin-dspace#1194)
remove duplicate dependency element with identical content (warnings in build)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added methods for counting handles and finding handles by token with enhanced validation.
  - Introduced a random string generator component to improve handle creation reliability.
  - Enhanced Clarin License confirmation handling using a new enum for clearer state representation.

- **Bug Fixes**
  - Improved error responses and logging for handle operations and license pagination.
  - Fixed regex pattern matching in handle validation and improved null safety in bitstream processing.

- **Refactor**
  - Replaced raw integer license confirmation values with descriptive enum constants across code and tests.
  - Standardized exception handling by replacing generic bad request exceptions with a custom application-specific exception.
  - Improved error logging strategy to differentiate client and server errors with configurable logging levels.
  - Updated handle URL generation to include fresh tokens and improved authorization handling during updates.

- **Tests**
  - Expanded integration tests for handle management and Clarin License workflows, including token validation and blacklist enforcement.
  - Added tests for pagination edge cases and unauthorized access scenarios based on license confirmation states.
  - Enhanced test setup and cleanup to maintain environment consistency.

- **Chores**
  - Removed duplicate dependencies and unused imports.
  - Updated configuration to enforce stricter URL and host validation rules for handle shortener functionality.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

OyvindLGjesdal committed:
remove duplicate dependency element with identical content (warnings in build)

(cherry picked from commit 59cee2721b88b5810c1f7ec5af612c0dc4ac8ca7)

kuchtiak-ufal committed:
Issue 1165: Handle PaginantionException in Clarin licences search requests (ufal/clarin-dspace#1184)

* Issue 1165: Handle PaginantionException in Clarin licences search requests

* resolve MR comments

---------

Co-authored-by: Milan Kuchtiak <mkuchtiak@Milans-MacBook-Pro.local>
(cherry picked from commit 1778714d3f2a832f571095a95ef0c1a3018bea53)

kuchtiak-ufal committed:
Issue 1055: fixed security issue for downloading file with non anonymous license (ufal/clarin-dspace#1188)

* Issue 1055: fixed security issue for downloading file with a license that requires authenticated user, also change ClarinLicense confirmation type from Integer to Enum (this reqired some refactoring)

* fixed failing tests

* add unit test for not authenticated user trying to get token for licence not allowed for anonymous user

* inmprove unit test

* resolve MR comments

* typo: backwards compatibility -> backward compatibility

---------

Co-authored-by: Milan Kuchtiak <mkuchtiak@Milans-MacBook-Pro.local>
(cherry picked from commit 53e88cb616471f83a28a9def37ec04ad0a30e036)

kuchtiak-ufal committed:
Issue 1186: catch DSpaceBadRequestException rather than BadRequestException (ufal/clarin-dspace#1187)

* Issue 1186: catch DSpaceBadRequestException rather than BadRequestException

* resolve test failures

* improve javadoc

* resolve MR comment

---------

Co-authored-by: Milan Kuchtiak <mkuchtiak@Milans-MacBook-Pro.local>
Co-authored-by: Ondřej Košarko <kosarko@ufal.mff.cuni.cz>
(cherry picked from commit 77fc4b057ef45e3520d85dc076134973ad78f476)

kuchtiak-ufal committed:
Issue ufal/clarin-dspace#1190 - fixed Server error  (ufal/clarin-dspace#1193)

* Issue 1190: fixed internal server error, when context.commit() was called multiple times in search request

* Issue 1190: fixed internal server error, when context.commit() was called multiple times in search request

* unnecessary whitespace

* update unit test

* small improvements

---------

Co-authored-by: Milan Kuchtiak <mkuchtiak@Milans-MacBook-Pro.local>
(cherry picked from commit 6638293b348d76ed49af3b7433cbeb0d2414e3cb)

kosarko committed:
Shortref v7 (ufal/clarin-dspace#1194)

* Anonymous users should be able to use shortener

* Update possible only with a valid token

* updated handle returned with usable token

additional cleanup - magicurl is an implementation detail and should not
be serialized

* checkstyle

* createHandle would throw ConstraintViolation

update the tests to use shortener.handle.prefix
handle validation - subprefix has a default, should not be required

* test blacklist/whitelist

checkstyle and license

* Address PR review comments

doc + default

(cherry picked from commit 2e3bd80bc02ac308773b4f7eec85a4b564fa6edc)

milanmajchrak commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/RandomStringGeneratorImpl.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package org.dspace.app.rest.repository;
+
+import org.apache.commons.lang.RandomStringUtils;
```
Following the code rabit review:
```
The code is using the older, deprecated version of Apache Commons Lang. This version is no longer maintained and may have security vulnerabilities.

-import org.apache.commons.lang.RandomStringUtils;
+import org.apache.commons.lang3.RandomStringUtils;

Consider updating the dependency in your build configuration to use the newer version.
```

kosarko committed:
Pull request review comment

use a recent version of commons-lang

vidiecan commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/ExternalHandleRestRepository.java:
```diff
@@ -257,7 +279,7 @@ private boolean matchesAnyOf(String tested, String configPropertyWithPatterns) {
         }
 
-        final String pattern = String.join(",", patterns);
+        final String pattern = String.join(";", patterns);
```
This is probably not correct:
shortener.post.url.blacklist.regexps = .*://.{1,15}

matchesAnyOf(host, "shortener.post.host.blacklist.regexps")

getArrayProperty splits by `,` you join it with `;` and split again with `;`

kosarko replied:
@vidiecan 
see the hostBlacklist test, I'm hoping that one is testing the config correctly.

getArrayProperty works in two "modes"
```
a=1
a=2
a=3
```
and 
```
a=1,2,3
```
and these should produce the same result `["1", "2", "3"]`.

In v5 we've used `;` for a similar purpose - to define an array of patterns. I'm assuming we are keeping that in v7 for compatibility. 

But what happens if you mix the two approaches?

```
a=1
a=2;3
```
getArrayProperty will load that as `["1", "2;3"]`? If you join that with `,` and then split with `;` you get `["1,2", "3"]`, but I'd say the correct result is `["1", "2", "3"]` and you get that with this changes, at least I hope you do.

kosarko replied:
>shortener.post.url.blacklist.regexps = .*://.{1,15}

I see, it's broken when there's a comma in the regexep

kosarko replied:
@vidiecan Seems that the config value is always split on commas (no difference between `getArrayProperty` and `getProperty`), so the reasonable thing to do here is escape the commas where you don't want the split to happen. I've added an example of that into clarin-dspace.cfg.

Also this is probably a lindat only feature and currently we are (in production) blacklisting everything but 3 domains.

vidiecan commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:
```diff
@@ -87,15 +86,15 @@ public Page<MetadataBitstreamWrapperRest> findByHandle(@Parameter(value = "handl
                 Thumbnail thumbnail = itemService.getThumbnail(context, item, false);
                 if (Objects.nonNull(thumbnail)) {
-                    bitstreams = new ArrayList<>();
+                    bitstreams.clear();
```
what exactly should this do?

kuchtiak-ufal replied:
This is just refactoring, due to some warning (in IntelliG) about reassigning local variables.

Instead of
```
List<Bitstream> bitstreams = bundle.getBitstreams();
if (condition) {
   bitstream = new ArrayList();
}
```

There is now:
```
 List<Bitstream> bitstreams = new ArrayList<>(bundle.getBitstreams());
 if (condition) {
    bitstream.clear();
 }
```

There is also no danger now that if something happens with **bitstreams** list - that would influence **bundle.bitstreams** list;

vidiecan commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/MetadataBitstreamRestRepository.java:
```diff
@@ -109,32 +108,30 @@ public Page<MetadataBitstreamWrapperRest> findByHandle(@Parameter(value = "handl
                         // Generate new content if we didn't find any
                         if (prContents.isEmpty()) {
-                            fileInfos = previewContentService.getFilePreviewContent(context, bitstream);
+                            fileInfos.addAll(previewContentService.getFilePreviewContent(context, bitstream));
```
why this change?

kuchtiak-ufal replied:
Similar as above -> avoid reassigning local variables.

kosarko committed:
Merge remote-tracking branch 'dataquest-dev/dtq-dev' into fork/ufal/ufal-dtq-sync

kosarko committed:
Shortener blacklist test

- provide more guidance in the config file
- change the default to match lindat prod

vidiecan commented on dspace-server-webapp/pom.xml:
```diff
@@ -290,12 +290,6 @@
             <version>${spring-boot.version}</version>
         </dependency>
 
-        <dependency>
```
are we sure about this?

what about

https://github.com/dataquest-dev/DSpace/blob/6daa0fdd88e4d377b16c01ec70430c68b89bb06e/dspace-server-webapp/src/main/java/org/dspace/app/rest/configuration/ActuatorConfiguration.java#L23

kosarko replied:
@vidiecan yes, see the lines above this: L287-L291
this is originally from https://github.com/ufal/clarin-dspace/pull/1138/files
